book/scheduling-and-threading.md

```diff
@@ -93,137 +43,139 @@ class Task:
         self.args = args
         self.__name__ = "task"
 
-    def __call__(self):
+    def run(self):
         self.task_code(*self.args)
         self.task_code = None
         self.args = None
 ```
 
-Also define a new `TaskRunner` class to manage the task queues and run them. It
-will have a list of `Task`s, a method to add a `Task`, and a method to run once
-through the event loop. Implement a simple scheduling heuristic in
-`run` that executes one task each time through, if there is one to run.
+The point of a task is that it can be created at one point in time,
+and then run at some later time by a task runner of some kind,
+according to a scheduling algorithm.[^event-loop] In our browser, the
+task runner will store tasks in a first-in first-out queue:
 
-``` {.python expected=False}
+[^event-loop]: The event loops we discussed in [Chapter
+2](graphics.md#eventloop) and [Chapter
+11](visual-effects.md#sdl-creates-the-window) are task runners, where
+the tasks to run are provided by the operating system.
+
+``` {.python replace=(self)/(self%2c%20tab)}
 class TaskRunner:
     def __init__(self):
         self.tasks = []
 
     def schedule_task(self, callback):
         self.tasks.append(callback)
+```
+
+When the time comes to run a task, our task runner can just remove
+the first task from the queue and run it:
 
+``` {.python expected=False}
+class TaskRunner:
     def run(self):
         if len(self.tasks) > 0):
             task = self.tasks.pop(0)
-            task()
+            task.run()
+
+class Tab:
+    def __init__(self):
+        self.task_runner = TaskRunner()
 ```
 
-Finally we just need to modify the main event loop to run the task runner each
-time, in case there is a script task to execute:
+First-in-first-out is a simplistic way to choose which task to run
+next, and real browsers have sophisticated *schedulers* which consider
+things like XXX.
```

Comment:
```suggestion
many different factors.
```

Reply:
There is also this link, but it's kind of old and weak sauce:

Reply:
https://blog.chromium.org/2015/04/scheduling-tasks-intelligently-for_30.html

Reply:
I'll include the link, better than nothing.